### ISSUE TITLE
Jedi does not work with C extension packages

### ISSUE THREAD

**Issue by jmcmorris** [davidhalter/jedi#609]
After a day of hacking around with Jedi I managed to get my siege.pyd package to be correctly picked up by Jedi; however, if I tried to get a submodule (siege.util) then it would completely fail.

Digging into `jedi.evaluate.imports.Importer._do_import()` I discovered that it correctly finds siege module but when it is working recursively `if len(import_path) > 1:` it fails. This is because the base for siege.util is siege which is a `jedi.evaluate.compiled.CompiledObject` and it does not have a `py__path__()`. Even if it did this would not work because later `jedi._compatibility.find_module()` is called which uses `imp.find_module()` which needs a path and doesn't work with hierarchical module names.

With that all being said, I managed to hack Jedi to work for this use case by checking if the base is a `CompiledObject` and if so then using `__import__()` and creating another `CompiledObject`. I do this directly after `base = bases[0]` in `jedi.evaluate.imports.Importer._do_import()`.

```python
if isinstance(base, compiled.CompiledObject):
    module = compiled.CompiledObject(__import__(module_name, fromlist=[base.obj]))
    return [module]
```

As I said this is a hack and I'm sure there is a much more elegant way to solve this; however, the Force is not strong with me yet and I've only begun to understand how Jedi works. With that being said, I do hope that a fix for this is implemented because I'd love to share my siege.pyd package with others in the future and be able to recommend Jedi to them.

Thanks for such an awesome Python tool!

**Comment by davidhalter** [comment]
Hi @jmcmorris 
Thanks for the detailed explanation! :)

What Python version are you testing this one against?

I'm not really sure why your module is not working, because basically the way the import logic of jedi is pretty much a reimplementation of Python's `importlib` (with similar names, usually prefixed by `py__XXX__`. That doesn't mean your implementation is wrong, Jedi probably just hasn't implemented a certain feature of the import logic. 

Is there a simple test case that you could provide? Also if you're interested in researching this: Why is `_init_module_attrs` (an importlib function that creates the `__path__` attribute) in your module  not called? 

May the force be with you ;-)

**Comment by jmcmorris** [comment]
Hey @davidhalter, thanks for the quick reply!

I'm using Python 2.7.5 atm (didn't realize I wasn't on 2.7.10!). Unfortunately importlib is really only for 3.1+ from what I understand which is why `_init_module_attrs` is not being called.

Here is the siege.pyd file I mentioned: https://www.dropbox.com/s/bg0txg7jjhroula/siege.zip?dl=0

Sorry for the size, this package is quite huge because it is actually an entire game engine exposed to python and along with it are all of the dependencies it has. If you try to import siege from outside of the cwd then you'll first need to add an environment variable, SIEGE_PATH that points to the directory that you extracted that zip - I could not for the life of me figure out how to get the module path from inside of C.

Let me know if you have any questions or anything.

**Comment by davidhalter** [comment]
Hmm well the question then is: How are we going to handle `__path__` in case of compiled modules for Python 2.6-3.2? - is it even different?

There are two things that we have to research:

1. When and in which Python version was `__path__` added to Python? Has it even existed before Python 3? (Which I think it does)
2. In case it existed: How did the import logic work if no `__path__` was provided. Please also provide the source of where in cPython that happens.

I won't do this research, since I have already done it for importlib and it was complicated there already. If someone wants to do it for `import.c` in Python 2, I would be really glad.

**Comment by jmcmorris** [comment]
I've spent the morning digging into this and here's what little I did come up with.

`__path__` is available in Python 2 as seen here: https://docs.python.org/2/tutorial/modules.html#packages-in-multiple-directories

From what I can tell though extensions packages are not given a __path__. I may be wrong with this but I can't find anything connecting the two.

I dug into `import.c` and focused on `find_module`. From what I can tell, `__path__` is not used at all and all possible attempts at providing a path resulted in "No frozen submodule named ..." which is the same error you get if you try to import a submodule for a module path that hasn't been imported yet.

Python 2.7.10:
`import.c` http://pastebin.com/kRUjSVyT
`importdl.c` http://pastebin.com/qKUjSdq8

I think the more interesting bit is that siege.pyd is considered a `imp.C_EXTENSION` which, when using `imp.load_module()` will end up calling `_PyImport_LoadDynamicModule`. This has no mention of `__path__` only `load_package()` does which is called from `load_module()` but that's only called for modules of type `imp.PKG_DIRECTORY`.

This is the most I could gleam from it for now. I've never dug into Python source before so this is all very unfamiliar territory to me. I'm considering posting in stackoverflow.com about this but wanted to first post here.

**Comment by davidhalter** [comment]
> I dug into import.c and focused on find_module

I'm not sure if that's important for our problem. My question is basically if there's no path (`__path__`), how does it even work? When you import `from siege import util`, then the following happens:

1. `sys.modules` is checked for `siege.util` returned if availble otherwise the next steps happen:
2. `sys.modules` is checked for `siege`. If not available that's imported, if available fine.
3. Now the tricky part: you want to import `util` and you have the package `siege`. How does Python's import logic know where it should load `util` from? It must be something that is defined in `siege`. It's usually `__path__`, even in Python2.7. I just checked this:

```
python2.7 -c "import json; print(json.__path__)"
['/usr/lib/python2.7/json']
```
----

I also just remembered that it's pretty typical that you have `__path__` defined on a module, because namespace packages in python 2.7 are using it.
